src/server/conn.rs

```diff
@@ -168,68 +152,70 @@ impl Conn {
               S: StoreAddrResolver
     {
         // all content should be read, ignore any read operation.
-        if self.payload.is_none() {
+        if self.recv_buffer.is_none() {
             return Ok(());
         }
         // TODO: limit rate
-        while try!(self.read_payload()) {
-            let payload = self.payload.take().unwrap();
-            let cap = payload.capacity();
-            self.read_size += cap;
+        loop {
+            {
+                let recv_buffer = self.recv_buffer.as_mut().unwrap();
+                try!(recv_buffer.read_from(&mut self.sock));
+                // if the snapshot is too small, the default buffer may be not filled.
+                if !recv_buffer.is_full() && recv_buffer.len() < self.expect_size {
+                    break;
+                }
+                if recv_buffer.len() > self.expect_size {
+                    return Err(box_err!("recv too much data!"));
+                }
+            }
+            let recv_buffer = self.recv_buffer.take().unwrap();
+            self.expect_size -= recv_buffer.len();
 
-            let task = SnapTask::Write(self.token, payload.flip());
+            let task = SnapTask::Write(self.token, recv_buffer);
             box_try!(self.snap_scheduler.schedule(task));
 
-            if self.read_size == self.file_size {
+            if self.expect_size == 0 {
                 // last chunk
                 box_try!(self.snap_scheduler.schedule(SnapTask::Close(self.token)));
                 // let snap_scheduler to close the connection.
                 break;
-            } else if self.read_size + cap >= self.file_size {
-                self.payload = Some(create_mem_buf(self.file_size - self.read_size))
+            } else if SNAPSHOT_PAYLOAD_BUF >= self.expect_size {
+                self.recv_buffer = Some(PipeBuffer::new(self.expect_size));
```

Comment:
recv_buffer need to be written to file in another thread, reuse it will have to clone the buffer, which is unnecessary.